File: Hadoop_Map_reduce_code/join/src/main/java/hadoop/join/RatingUsers.java
package hadoop.join;

import java.net.URI;
import java.util.HashMap;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapreduce.Job;
import org.apache.hadoop.mapreduce.lib.input.MultipleInputs;
import org.apache.hadoop.mapreduce.lib.input.TextInputFormat;
import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat;

public class RatingUsers {

	public static void main(String[] args) throws Exception{

		Configuration conf = new Configuration();
		
		HashMap<String, Integer> mapRating = new HashMap<>(); 
		mapRating.put("userid",0);
		mapRating.put("movieid",1);
		mapRating.put("rating",2);
		mapRating.put("timestamp",3);
		
		HashMap<String, Integer> mapUsers = new HashMap<>(); 
		mapUsers.put("userid",0);
		mapUsers.put("age",1);
		mapUsers.put("gender",2);
		mapUsers.put("occupation",3);
		mapUsers.put("zipcode",4);
		
		String attribute="DEFAULT";
		String val = "DEFAULT";
		if(args.length==2)
		{
			attribute=args[0];
			val=args[1];
		}
		
		if(mapRating.containsKey(attribute))
		{
			conf.set("table1","rating");
			conf.set("attributeIndex1", String.valueOf(mapRating.get(attribute)));
			conf.set("val1", val);
		}
		
		if(mapUsers.containsKey(attribute))
		{
			conf.set("table2","users");
			conf.set("attributeIndex2", String.valueOf(mapUsers.get(attribute)));
			conf.set("val2", val);
		}
		
		Job job = new Job(conf, "Rating Users");
		 job.setJarByClass(RatingUsers.class);
		 job.setReducerClass(RatingUsersReducer.class);
		 job.setOutputKeyClass(Text.class);
		 job.setOutputValueClass(Text.class);
		 
		 
		 
		 MultipleInputs.addInputPath(job, new Path("/cloud/input/rating.csv"),TextInputFormat.class, RatingMapperRU.class);
		 MultipleInputs.addInputPath(job, new Path("/cloud/input/users.csv"),TextInputFormat.class, UsersMapperRU.class);
		 Path outputPath = new Path("/cloud/output/ratingusers");
		  
		 FileOutputFormat.setOutputPath(job, outputPath);
		 
		 FileSystem hdfs = FileSystem.get(URI.create("hdfs://localhost:9000"),conf);
			
			// delete existing directory
			if (hdfs.exists(outputPath)) {
			    hdfs.delete(outputPath, true);
			}
		 
		 
		 
		 
		 System.exit(job.waitForCompletion(true) ? 0 : 1);
		
	}

}

File: spark_scripts/ratingusers.py
from pyspark import SparkContext, SparkConf
import sys
import time

start_time = time.time()

conf = SparkConf().setAppName("ratingusers")
sc = SparkContext(conf = conf)

URI = sc._gateway.jvm.java.net.URI
Path = sc._gateway.jvm.org.apache.hadoop.fs.Path
FileSystem = sc._gateway.jvm.org.apache.hadoop.fs.FileSystem
fs = FileSystem.get(URI("hdfs://localhost:9000"), sc._jsc.hadoopConfiguration())


ratingRDD = sc.textFile("hdfs://localhost:9000/cloud/input/rating.csv")
ratingRDDMap = ratingRDD.map(lambda rec: (rec.split(",")[0], rec))

usersRDD = sc.textFile("hdfs://localhost:9000/cloud/input/users.csv")
usersRDDMap = usersRDD.map(lambda rec: (rec.split(",")[0], rec))

ratingusersRDDMap=ratingRDDMap.join(usersRDDMap)
ratingusersRDDMapF = ratingusersRDDMap

mapRating = {
  "userid": 0,
  "movieid": 1,
  "rating": 2,
  "timestamp": 3,
}

mapUsers =	{
  "userid": 0,
  "age": 1,
  "gender": 2,
  "occupation": 3,
  "zipcode": 4,
}



def main():
  arguments = sys.argv[1:]
  count = len(arguments)
  if count == 2:
    attribute = arguments[0]
    val = arguments[1]
    attributeIndexRating = mapRating.get(attribute)
    attributeIndexUsers = mapUsers.get(attribute)
    if attributeIndexRating != None:
      ratingusersRDDMapF = ratingusersRDDMap.filter(lambda row: int(row[1][0].split(",")[attributeIndexRating]) == int(val))
    elif attributeIndexUsers != None:
      if attribute == 'userid' or attribute == 'age':
        ratingusersRDDMapF = ratingusersRDDMap.filter(lambda row: int(row[1][1].split(",")[attributeIndexUsers]) == int(val))
      else:
        ratingusersRDDMapF = ratingusersRDDMap.filter(lambda row: (val in row[1][1].split(",")[attributeIndexUsers]))
    else:
      ratingusersRDDMapF = ratingusersRDDMap
  else:
    ratingusersRDDMapF = ratingusersRDDMap
  
  #for i in ratingusersRDDMapF.collect():
  #  print(i)
  
  fs.delete(Path('/cloud/output/spark/ratingusers') ,True)
  ratingusersRDDMapF.coalesce(1, shuffle = True).saveAsTextFile("hdfs://localhost:9000/cloud/output/spark/ratingusers")

  execution_time = time.time() - start_time
  f = open("extime.txt", "w")
  f.write(str(execution_time))
  f.close()  

if __name__== "__main__":
  main()
